### PR TITLE
Boolean Hamiltonian gate

### PR DESCRIPTION
**Pull request by tonybruguier**
Following #4282, the present PR would like to add a gate that allows computing the Hamiltonian from a Boolean expression. Note that while the decomposition is somewhat efficient, more optimizations are possible and planned in a follow-up PR.

### CODE REVIEW

**Comment by tonybruguier** [comment]
I would suggest @balopat for review (this message should not be taken as a ping)

**Comment by tonybruguier** [comment]
Friendly ping to @balopat for review. This has largely been reviewed by you already except that I decided to spilt into a smaller PR that I think has the minimum for the gate to function.

File: cirq-core/cirq/ops/boolean_hamiltonian.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        theta: float,
+    ):
+        """Builds a BooleanHamiltonian.
+
+        For each element of a sequence of Boolean expressions, the code first transforms it into a
+        polynomial of Pauli Zs that represent that particular expression. Then, we sum all the
+        polynomials, thus making a function that goes from a series to Boolean inputs to an integer
+        that is the number of Boolean expressions that are true.
+
+        For example, if we were using this gate for the max-cut problem that is typically used to
```
Comment (balopat) [inline cirq-core/cirq/ops/boolean_hamiltonian.py:51]:
```suggestion
        For example, if we were using this gate for the unweighted max-cut problem that is typically used to
```

**Review by balopat — APPROVED** [review]
I think this looks good, just found a couple of nits, please add those.

**Comment by tonybruguier** [comment]
Thanks @balopat !
All the suggestions made sense, so I accepted them, merged the main branch, and re-ran the tests.
PTAL.

**Pull request merged**